src/backend/web.rs

```diff
@@ -55,19 +54,46 @@ pub(crate) struct RenderBundleEncoder(web_sys::GpuRenderBundleEncoder);
 // This is safe on wasm32 *for now*, but similarly to the unsafe Send impls for the handle type
 // wrappers, the full story for threading on wasm32 is still unfolding.
 
-pub(crate) struct MakeSendFuture<F>(F);
+pub(crate) struct MakeSendFuture<F, M> {
+    future: F,
+    map: Option<M>,
+}
+
+impl<F: Future, M: FnOnce(F::Output) -> T, T> Future for MakeSendFuture<F, M> {
+    type Output = T;
 
-impl<F: Future> Future for MakeSendFuture<F> {
-    type Output = F::Output;
+    fn poll(mut self: Pin<&mut Self>, cx: &mut task::Context) -> Poll<Self::Output> {
+        assert!(
+            self.as_mut().get_mut().map.is_some(),
+            "MakeSendFuture must not be polled after returning Ready"
+        );
 
-    fn poll(self: Pin<&mut Self>, cx: &mut task::Context) -> Poll<Self::Output> {
         // This is safe because we have no Drop implementation to violate the Pin requirements and
         // do not provide any means of moving the inner future.
-        unsafe { self.map_unchecked_mut(|s| &mut s.0) }.poll(cx)
+        let future = unsafe { self.as_mut().map_unchecked_mut(|s| &mut s.future) };
```

Comment:
We could probably simplify the access to `T` a little, e.g.:
```rust
fn poll(self: Pin<&mut Self>, cx: &mut task::Context) -> Poll<Self::Output> {
    // This is safe because we have no Drop implementation to violate the Pin requirements and
    // do not provide any means of moving the inner future.
    unsafe {
        let this = self.get_unchecked_mut();
        match Pin::new_unchecked(&mut this.future).poll(cx) {
            task::Poll::Ready(value) => task::Poll::Ready((this.map)(value)),
            task::Poll::Pending => task::Poll::Pending,
        }
    }
}
```